autorest/azure/environments.go

```diff
@@ -0,0 +1,99 @@
+		GalleryEndpoint:           "https://gallery.azure.com/",
+		KeyVaultEndpoint:          "https://vault.azure.net/",
+		GraphEndpoint:             "https://graph.windows.net/",
+		StorageEndpointSuffix:     "core.windows.net",
+		SQLDatabaseDNSSuffix:      "database.windows.net",
+		TrafficManagerDNSSuffix:   "trafficmanager.net",
+		KeyVaultDNSSuffix:         "vault.azure.net",
+		ServiceBusEndpointSuffix:  "servicebus.azure.com",
+	}
+
+	// USGovernmentCloud is the cloud environment for the US Government
+	USGovernmentCloud = Environment{
+		Name:                      "AzureUSGovernmentCloud",
+		ManagementPortalURL:       "https://manage.windowsazure.us/",
+		PublishSettingsURL:        "https://manage.windowsazure.us/publishsettings/index",
+		ServiceManagementEndpoint: "https://management.core.usgovcloudapi.net/",
+		ResourceManagerEndpoint:   "https://management.usgovcloudapi.net",
+		ActiveDirectoryEndpoint:   "https://login.microsoftonline.com/?api-version=1.0",
+		GalleryEndpoint:           "https://gallery.usgovcloudapi.net/",
+		KeyVaultEndpoint:          "https://vault.azure.net/",
+		GraphEndpoint:             "https://graph.usgovcloudapi.net/",
+		StorageEndpointSuffix:     "core.usgovcloudapi.net",
+		SQLDatabaseDNSSuffix:      "database.usgovcloudapi.net",
+		TrafficManagerDNSSuffix:   "trafficmanager.net",
+		KeyVaultDNSSuffix:         "vault.azure.net",
+		ServiceBusEndpointSuffix:  "servicebus.usgovcloudapi.net",
+	}
+
+	// ChinaCloud is the cloud environment operated in China
+	ChinaCloud = Environment{
+		Name:                      "AzureChinaCloud",
+		ManagementPortalURL:       "https://manage.chinacloudapi.com/",
+		PublishSettingsURL:        "https://manage.chinacloudapi.com/publishsettings/index",
+		ServiceManagementEndpoint: "https://management.core.chinacloudapi.cn/",
+		ResourceManagerEndpoint:   "https://management.chinacloudapi.cn/",
+		ActiveDirectoryEndpoint:   "https://login.chinacloudapi.cn/?api-version=1.0",
+		GalleryEndpoint:           "https://gallery.chinacloudapi.cn/",
+		KeyVaultEndpoint:          "https://vault.azure.net/",
+		GraphEndpoint:             "https://graph.chinacloudapi.cn/",
+		StorageEndpointSuffix:     "core.chinacloudapi.cn",
+		SQLDatabaseDNSSuffix:      "database.chinacloudapi.cn",
+		TrafficManagerDNSSuffix:   "trafficmanager.cn",
+		KeyVaultDNSSuffix:         "vault.azure.net",
+		ServiceBusEndpointSuffix:  "servicebus.chinacloudapi.net",
+	}
+)
+
+// OAuthConfigForTenant returns an OAuthConfig wired for the given tenant
+func (env Environment) OAuthConfigForTenant(tenantID string) (*OAuthConfig, error) {
+	parsedURL, err := url.Parse(env.ActiveDirectoryEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to construct oauth configuration. Error parsing base URL: %q", err)
+	}
+	authorizeURL, tokenURL, deviceCodeURL := *parsedURL, *parsedURL, *parsedURL
+	authorizeURL.Path = path.Join(authorizeURL.Path, tenantID, "oauth2", "authorize")
```

Comment:
Agreed, fixed. I didn't realize that url.URL had it's own Parse method. Definitely is cleaner now.